qiskit/circuit/classical/expr/expr.py

```diff
@@ -0,0 +1,288 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Expression-tree nodes."""
+
+# Given the nature of the tree representation and that there are helper functions associated with
+# many of the classes whose arguments naturally share names with themselves, it's inconvenient to
+# use synonyms everywhere.  This goes for the builtin 'type' as well.
+# pylint: disable=redefined-builtin,redefined-outer-name
+
+from __future__ import annotations
+
+__all__ = [
+    "ExprVisitor",
+    "Expr",
+    "Value",
+    "Cast",
+    "Unary",
+    "Binary",
+]
+
+import abc
+import enum
+import typing
+
+from .. import types
+
+if typing.TYPE_CHECKING:
+    import qiskit
+
+_T_co = typing.TypeVar("_T_co", covariant=True)
+
+
+class ExprVisitor(typing.Generic[_T_co]):
+    """Base class for visitors to the :class:`Expr` tree.  Subclasses should override whichever of
+    the ``visit_*`` methods that they are able to handle, and should be organised such that
+    non-existent methods will never be called."""
+
+    # pylint: disable=missing-function-docstring
+
+    __slots__ = ()
+
+    def visit_generic(self, expr: "Expr", /) -> _T_co:  # pragma: no cover
+        raise RuntimeError(f"expression visitor {self} has no method to handle expr {expr}")
+
+    def visit_value(self, expr: "Value", /) -> _T_co:  # pragma: no cover
+        return self.visit_generic(expr)
+
+    def visit_unary(self, expr: "Unary", /) -> _T_co:  # pragma: no cover
+        return self.visit_generic(expr)
+
+    def visit_binary(self, expr: "Binary", /) -> _T_co:  # pragma: no cover
+        return self.visit_generic(expr)
+
+    def visit_cast(self, expr: "Cast", /) -> _T_co:  # pragma: no cover
+        return self.visit_generic(expr)
+
+
+class Expr(abc.ABC):
```

Comment:
At some point we likely want to explore uniting the *full* Qiskit program representation within an object model (ie., unite Expr, Circuit, Operations, Pulses, etc) within a common object model that has common properties. Each component being different makes it difficult to operate/extend in a unified fashion as the complexity grows.

I understand this is outside the scope of this PR, but it has to be tackled at *some point*.

Reply:
I know that say Clang has its `Expr` class inherit from `ValueStatement`, so they unify those more strongly than our initial plans.  With how Terra is likely to reason (or not) about classical operations, I'm not certain that that's going to be the best strategy for us.

I think our representation of classical processing is likely to (certainly at the start) be more along the lines of just "store this value" and "use this value", and leave the details of lowering/optimising the classical components to lower parts of the stack.  We can define the evaluation order of arguments, but then if Terra's not actually going to be doing any reasoning about how classical computations are lowered to hardware, then I'm not certain how much benefit there is in having every `Expr` be required to be linked one-to-one with a "statement".